### PR TITLE
Attempt to reduce code 500 errors

### PR DESCRIPTION
**Pull request by russgold**
This addresses issue 155, which reports a NPE in WLS. WLS is clearly not testing the message parameter for null, so we have to ensure that we never send a null.

It also attempts to address a problem when the public DNS name of the server is not locally resolvable (a very odd case) by trying "localhost" as an alternative.

### CODE REVIEW

**Review by ddsharpe — APPROVED** [review]

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java [hunk 1]
```diff
@@ -68,7 +70,11 @@ public PrintStream getResponseStream() throws IOException {
 
   @Override
   public void sendError(int status, String msg) throws IOException {
+    if (msg == null) {
```
Comment (alai8) [inline wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java:73]: was the NPE caused by a null msg value?

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/NotFoundException.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2021, Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+package com.oracle.wls.exporter;
+
+/**
+ * An exception indicating that the client could not connect to the specified URL.
+ */
+public class NotFoundException extends WebClientException {
```
Comment (alai8) [inline wls-exporter-core/src/main/java/com/oracle/wls/exporter/NotFoundException.java:9]: I can't seem to find any usages of this new class

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/UrlBuilder.java [hunk 3]
```diff
@@ -20,26 +22,48 @@
  */
 public class UrlBuilder {
 
-  private static final Set<Integer> successes = new ConcurrentSkipListSet<>();
+  private static class WebHost implements Comparable<WebHost> {
+    private final String hostName;
+    private final int port;
+
+    public WebHost(String hostName, int port) {
```
Comment (alai8) [inline wls-exporter-core/src/main/java/com/oracle/wls/exporter/UrlBuilder.java:29]: would it be better to name this inner class such as WebHostPort?

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java [hunk 4]
```diff
@@ -31,7 +31,9 @@ public void close() {
 
   @Override
   public UrlBuilder createUrlBuilder() {
-    return UrlBuilder.create(request.getLocalName(), request.isSecure())
+    return UrlBuilder.create(request.isSecure())
+          .withHostName(request.getLocalName())
+          .withHostName("localhost")
```
Comment (alai8) [inline wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java:36]: if I understand correctly, we will always try localhost if the connection using request.getLocalName() failed. Should this be documented somewhere?

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java (hunk 1 above)
Comment (russgold) [reply wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java:73]: yes

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/NotFoundException.java (hunk 2 above)
Comment (russgold) [reply wls-exporter-core/src/main/java/com/oracle/wls/exporter/NotFoundException.java:9]: Nice catch, thanks.

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/UrlBuilder.java (hunk 3 above)
Comment (russgold) [reply wls-exporter-core/src/main/java/com/oracle/wls/exporter/UrlBuilder.java:29]: I did some research into this. Apparently, to the extent that a standard exists, the name/port combination is called the "host" and the name alone is supposed to be called, "host name." So I renamed the variables appropriately. 

https://stackoverflow.com/questions/13673389/difference-between-host-and-hostname-in-window/13673475

File: wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java (hunk 4 above)
Comment (russgold) [reply wls-exporter-core/src/main/java/com/oracle/wls/exporter/ServletInvocationContext.java:36]: Yes, it should - and it should not be using a literal string. I will fix it.

**Review by ddsharpe — CHANGES_REQUESTED** [review]

**Review by alai8 — APPROVED** [review]

**Review by alai8 — APPROVED** [review]

**Review by ddsharpe — APPROVED** [review]

**Pull request merged**